composer/trainer/mosaic_fsdp.py

```diff
@@ -175,112 +173,122 @@ def get_process_group(pg, process_group_cache=None):
+            auto_wrap_policy (Callable):
+                A callable specifying a policy to recursively wrap layers with FSDP.
+            ignored_modules (Set[torch.nn.Module]): Modules to ignore when
+                wrapping.
+            ignored_params (Set[torch.nn.Parameter]): Parameters to ignore when
+                wrapping; these should be the parameters contained in the modules
+                in ``ignored_modules``.
+            process_group_cache (Dict[Tuple[int], Any]): a cache of process_group to
+                use instead of potentially instantiating a new process_group
+
+        Returns:
+            (nn.Module, int):
+                Wrapped module and the number parameters wrapped recursively.
+        """
+        assert auto_wrap_policy is not None, 'Must specify auto_wrap_policy.'
+        assert wrapper_cls is not None, 'Must specify wrapper_cls'
+        # Make sure no child is already wrapped.
+        for _, child in module.named_modules():
+            if child in ignored_modules:
+                continue
+            try:
+                assert not isinstance(child, cast(type, wrapper_cls))
+            except TypeError:
+                # wrapper_cls is a function as opposed to a class type, just bypass above check.
+                pass
 
-class MosaicFullyShardedDataParallel(FullyShardedDataParallel):
-    """Updates FSDP's _auto_wrap to enable module_kwargs."""
+        # We count all params, assuming none of them are already wrapped.
+        num_params = sum(p.numel() for p in module.parameters() if p not in ignored_params)
 
-    def _auto_wrap(
+        assert auto_wrap_policy is not None
+        if auto_wrap_policy(module=module, recurse=True, unwrapped_params=num_params):
+            total_wrapped_params = 0
+            # Iterate through the children, recursively wrap if necessary
+            for name, child in module.named_children():
+                if child in ignored_modules:
+                    continue
+                wrapped_child, num_wrapped_params = _custom_recursive_wrap_t1p13p1(
+                    module=child,
+                    auto_wrap_policy=auto_wrap_policy,
+                    wrapper_cls=wrapper_cls,
+                    ignored_modules=ignored_modules,
+                    ignored_params=ignored_params,
+                    process_group_cache=process_group_cache,
+                    **kwargs,
+                )
+                setattr(module, name, wrapped_child)
+                # Keep track of how many parameters have been wrapped
+                total_wrapped_params += num_wrapped_params
+            # decide if we need to wrap the current module,
+            # since the left over parameters exceed the number of params to wrap
+            remainder = num_params - total_wrapped_params
+            module_kwargs = auto_wrap_policy(module=module, recurse=False, unwrapped_params=remainder)
+            if not only_wrap_children and module_kwargs:
```

Comment:
valid
keeling both `_custom_recursive_wrap_t1p13p1` and `_custom_recursive_wrap_t2p0p1` but creating
```
def set_custom_fsdp_module_kwargs(module_kwargs: Dict) -> Dict:
```
since its the same in both versions